crates/standalone/src/lib.rs

```diff
@@ -289,51 +275,53 @@ impl spacetimedb_client_api::ControlStateWriteAccess for StandaloneEnv {
-            None => Database {
-                id: 0,
-                address: spec.address,
-                identity: *identity,
-                host_type: HostType::Wasm,
-                num_replicas: spec.num_replicas,
-                program_bytes_address,
-                publisher_address,
-            },
-        };
-
-        if let Some(existing) = existing_db.as_ref() {
-            anyhow::ensure!(
-                &existing.identity == identity,
-                "Permission denied: `{identity}` does not own database `{}`",
-                spec.address.to_abbreviated_hex()
-            );
-            self.control_db.update_database(database.clone())?;
-        } else {
-            let id = self.control_db.insert_database(database.clone())?;
-            database.id = id;
-        }
-
-        let database_id = database.id;
-        let should_update_instances = existing_db.is_some();
-
-        self.schedule_database(Some(database), existing_db).await?;
-
-        if should_update_instances {
-            let leader = self
-                .control_db
-                .get_leader_database_instance_by_database(database_id)
-                .ok_or_else(|| anyhow!("Not found: leader instance for database {database_id}"))?;
-            Ok(self.update_database_instance(leader).await?)
-        } else {
-            Ok(None)
+        match existing_db {
+            None => {
+                let program_bytes_address = self.object_db.insert_object(spec.program_bytes)?;
+                let mut database = Database {
+                    id: 0,
+                    address: spec.address,
+                    identity: *identity,
+                    // TODO: Add to `DatabaseDef`
+                    host_type: HostType::Wasm,
+                    num_replicas: spec.num_replicas,
+                    program_bytes_address,
+                    publisher_address,
+                };
+
+                let database_id = self.control_db.insert_database(database.clone())?;
+                database.id = database_id;
+
+                self.schedule_database(Some(database), None).await?;
+
+                Ok(None)
+            }
+            Some(existing_db) => {
+                ensure!(
+                    &existing_db.identity == identity,
+                    "Permission denied: `{}` does not own database `{}`",
+                    identity,
+                    spec.address.to_abbreviated_hex()
+                );
+
+                let program_bytes_address = self.object_db.insert_object(spec.program_bytes)?;
+                let database = Database {
+                    num_replicas: spec.num_replicas,
+                    program_bytes_address,
+                    publisher_address,
+                    ..existing_db.clone()
+                };
+                let database_id = database.id;
+                let database_addr = database.address;
+                self.control_db.update_database(database.clone())?;
```

Comment:
The logic did not change in this patch, but I'll point out that it is flawed:

If we fail to run the update, the new database will keep running.
There is no obvious solution for that within the current event-driven design, because it is impossible to guarantee that we can restore the previous state.

It will need to be revised in a later patch.